src/storage-controller/src/collection_mgmt.rs

```diff
@@ -1439,46 +1435,87 @@ where
+                    // so we don't bother being careful about it.
+                    //
+                    // TODO(guswynn): unpack these into health-status objects and use
+                    // their `Ord` impl.
+                    entries.push(Reverse((timestamp, row.clone())));
+
+                    // Retain some number of entries, using pop to mark the oldest entries for
+                    // deletion.
+                    while entries.len() > n {
+                        if let Some(Reverse((_, r))) = entries.pop() {
+                            deletions.push(r);
+                        }
+                    }
+                }
             }
         }
+        StatusHistoryRetentionPolicy::TimeWindow(time_window) => {
+            // Get the lower bound
+            let now = mz_ore::now::to_datetime(now());
+            let keep_since = now - time_window;
+
+            // Produce retractions by inverting diffs of rows we want to delete and setting the diffs
+            // of all other rows to 0.
+            for (row, diff) in rows {
+                let datums = row.unpack();
+                let key = (status_history_desc.extract_key)(&datums);
+                let timestamp = (status_history_desc.extract_time)(&datums);
+
+                assert!(
+                    diff > 0,
+                    "only know how to operate over consolidated data with diffs > 0, \
+                found diff {diff} for object {key:?} in {introspection_type:?}",
+                );
 
-        // Consider duplicated rows separately.
-        let entries = last_n_entries_per_key.entry(key).or_default();
-        for _ in 0..diff {
-            // We CAN have multiple statuses (most likely Starting and Running) at the exact same
-            // millisecond, depending on how the `health_operator` is scheduled.
-            //
-            // Note that these will be arbitrarily ordered, so a Starting event might
-            // survive and a Running one won't. The next restart will remove the other,
-            // so we don't bother being careful about it.
-            //
-            // TODO(guswynn): unpack these into health-status objects and use
-            // their `Ord` impl.
-            entries.push(Reverse((timestamp, row.clone())));
-
-            // Retain some number of entries, using pop to mark the oldest entries for
-            // deletion.
-            while entries.len() > status_history_desc.keep_n {
-                if let Some(Reverse((_, r))) = entries.pop() {
-                    deletions.push(r);
+                // Keep track of the timestamp of the latest row per key.
+                match latest_row_per_key.get(&key) {
+                    Some(existing) if &existing.0 > &timestamp => {}
+                    _ => {
+                        latest_row_per_key.insert(key.clone(), (timestamp, row.clone()));
+                    }
+                }
```

Comment:
This is a bit more duplicated code than I like to see. I think it _should_ work to extract a helper closure like this:

```
let handle_row = |row, diff| {
    let datums = row.unpack();
    let key = ...;

    // ...

    (key, timestamp)
};
```

And then you can put

```
let (key, timestamp) = handle_row(&row, diff);
```

into both arms of the match.